alternative registry support

tofay opened the pull request:
This PR adds support for alternative registries.

The main change is updating the `fetch::get_latest_dependency` to optionally accept a registry URL. I'm using URLs not registry names, because `cargo metadata` returns the URL of the index whenever the dependency is being sourced from an alternative registry.

When adding dependencies from alternative registries, the user specifies a registry name which is then converted to the relevant URL. For this purpose, the cargo config parsing code in `registry.rs` is updated to handle registries other than crates-io.

W.r.t testing, I've only added tests for the CLI/toml-editing aspects. I've been successfully using `cargo upgrade` with my company's alternative registry, but I've not added tests that actually query an alternative registry, because the tests currently rely on the user's cargo config.
Is that acceptable? (Cargo itself has private test fixtures for working with alternative registries - this could be reused when/if cargo-edit is moved into cargo.)

Fixes https://github.com/killercup/cargo-edit/issues/201
Fixes #339

ordian commented on src/bin/add/args.rs:
```diff
@@ -104,6 +104,15 @@ pub struct Args {
+        long = "registry",
+        conflicts_with = "git",
+        conflicts_with = "path",
+        conflicts_with = "build"
```
Are you sure it conflicts with build dependencies? Seems to work locally and I couldn't find it shouldn't in the RFC.
```suggestion
```

tofay replied:
I'm mistaken here, so will remove this, thanks!

ordian commented on src/bin/add/args.rs:
```diff
@@ -104,6 +104,15 @@ pub struct Args {
+    #[structopt(
+        long = "registry",
+        conflicts_with = "git",
+        conflicts_with = "path",
```
```suggestion
        conflicts_with = "path"
```

ordian commented on src/registry.rs:
```diff
@@ -38,13 +50,28 @@ fn cargo_home() -> Result<PathBuf> {
-pub fn registry_url(manifest_path: &Path) -> Result<Url> {
+/// Find the URL of a registry
+pub fn registry_url(manifest_path: &Path, registry: Option<&str>) -> Result<Url> {
+    if env::var("CARGO_IS_TEST").is_ok() {
```
why do we need this?

tofay replied:
I needed it to trick the `cargo add` tests into looking at crates-io, but it'll be unnecessary after I've added a non-default cargo config.

tofay commented:
Thanks for the review. I've rebased and added a test fixture that uses https://github.com/ordian/crates.io-index.

I had a couple of issues relating to cargo config:

- I hit https://github.com/killercup/cargo-edit/issues/339, and worked around it for now with a `cargo search` command.
- The cargo config probing differs slightly between cargo and cargo-edit. Cargo searches from the invocation directory whereas cargo-edit searches from the parent of the manifest path. The alternative registry cargo upgrade tests run commands from the temporary test directories themselves, to ensure that both the cargo metadata command and cargo-edit can detect the test cargo config.

tofay commented on src/dependency.rs:
```diff
@@ -109,6 +113,21 @@ impl Dependency {
 
+    /// Set the value of registry for the dependency
+    pub fn set_registry(mut self, registry: &str) -> Dependency {
+        let old_source = self.source;
```
I just c/p this from elsewhere in this file without thinking about it... It's not needed though, so I've removed it.

tofay commented on src/bin/upgrade/main.rs:
```diff
@@ -281,6 +303,21 @@ fn process(args: Args) -> Result<()> {
+            .0
+            .values()
+            .filter_map(|(registry, _)| registry.as_ref())
+            .unique()
```
Yes. I've removed the `itertools` dependency.

tofay commented on tests/utils.rs:
```diff
@@ -19,7 +19,29 @@ pub fn clone_out_test(source: &str) -> (tempdir::TempDir, String) {
+        path.join(".cargo").join("config"),
+    )
+    .unwrap_or_else(|err| panic!("could not copy test cargo config: {}", err));
+    // Workaround for https://github.com/killercup/cargo-edit/issues/339
```
Thanks! It suffices to init the repo instead of cloning it, so I've done that to avoid issues like https://github.com/killercup/cargo-edit/issues/333 when cloning as well as fetching.

tofay commented on src/bin/add/args.rs:
```diff
@@ -182,6 +194,12 @@ impl Args {
             }
 
+            // Set the registry after getting the latest version as
+            // get_latest_dependency returns a registry-less Dependency
```
It would complicate the code a bit more so I'll leave as is. (`get_latest_dependency` accepts a registry URL which both cargo-upgrade and cargo-add can easily provide. It's nontrivial for cargo-upgrade to provide a registry name.)
